In [1]:
#Import necessary libraries
import os
import pandas as pd
import json

In [2]:
#Set the working directory 
#os.chdir("/Users/Abdul/Desktop/MMA/Enterprise Data Science/Revenue-Radar/Data")


In [3]:
#Import dataset(s)
train_revenue = pd.read_csv("Data/train.csv")
test_revenue = pd.read_csv("Data/test.csv")

/var/folders/9z/gkk8fkkj0pz746hrzcw69dq40000gn/T/ipykernel_8409/2625106073.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  train_revenue = pd.read_csv("Data/train.csv")
/var/folders/9z/gkk8fkkj0pz746hrzcw69dq40000gn/T/ipykernel_8409/2625106073.py:3: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  test_revenue = pd.read_csv("Data/test.csv")


### Seprating the code from above to make the waiting time shorter and to have some progress

In [4]:
dict_columns = ['geoNetwork', 'device', 'totals', 'trafficSource']

# For each such column, convert the string representation of dictionary to actual dictionary
# then normalize and concatenate with the original DataFrame
for column in dict_columns:
    # Convert the stringified dictionary to actual dictionary
    train_revenue[column] = train_revenue[column].apply(json.loads)
    
    # Flatten the dictionary and convert to a DataFrame
    column_df = pd.json_normalize(train_revenue[column])
    
    # Concatenate the new DataFrame to the original DataFrame
    train_revenue = pd.concat([train_revenue, column_df], axis=1)

    # Drop the original column that contained the stringified dictionary
    train_revenue.drop(column, axis=1, inplace=True)



In [5]:
for column in dict_columns:
    # Convert the stringified dictionary to actual dictionary
    test_revenue[column] = test_revenue[column].apply(json.loads)
    
    # Flatten the dictionary and convert to a DataFrame
    column_df = pd.json_normalize(test_revenue[column])
    
    # Concatenate the new DataFrame to the original DataFrame
    test_revenue = pd.concat([test_revenue, column_df], axis=1)

    # Drop the original column that contained the stringified dictionary
    test_revenue.drop(column, axis=1, inplace=True)

### THE ABOVE TAKE TOOOO LONG to Run, let's save the data into new dataframe, avoid running it everytime
 

In [21]:
#  saving the data to new dataframes to make sure we don't have to run the above code again, it takes me 10 minutes to run
train_data = train_revenue
test_data = test_revenue

df_test = test_data
df_train = train_data

In [11]:
#Viewing first 5 rows of the training dataset
train_data.head()

channelGrouping       date        fullVisitorId  \
0  Organic Search 2016-09-02  1131660440785968503   
1  Organic Search 2016-09-02   377306020877927890   
2  Organic Search 2016-09-02  3895546263509774583   
3  Organic Search 2016-09-02  4763447161404445595   
4  Organic Search 2016-09-02    27294437909732085   

                        sessionId  socialEngagementType     visitId  \
0  1131660440785968503_1472830385  Not Socially Engaged  1472830385   
1   377306020877927890_1472880147  Not Socially Engaged  1472880147   
2  3895546263509774583_1472865386  Not Socially Engaged  1472865386   
3  4763447161404445595_1472881213  Not Socially Engaged  1472881213   
4    27294437909732085_1472822600  Not Socially Engaged  1472822600   

   visitNumber  visitStartTime continent     subContinent  ...  \
0            1      1472830385      Asia     Western Asia  ...   
1            1      1472880147   Oceania      Australasia  ...   
2            1      1472865386    Europe  Southern Europe  ...   
3            1      1472881213      Asia   Southeast Asia  ...   
4            2      1472822600    Europe  Northern Europe  ...   

  adwordsClickInfo.criteriaParameters isTrueDirect referralPath  \
0       not available in demo dataset          NaN          NaN   
1       not available in demo dataset          NaN          NaN   
2       not available in demo dataset          NaN          NaN   
3       not available in demo dataset          NaN          NaN   
4       not available in demo dataset         True          NaN   

  adwordsClickInfo.page adwordsClickInfo.slot adwordsClickInfo.gclId  \
0                   NaN                   NaN                    NaN   
1                   NaN                   NaN                    NaN   
2                   NaN                   NaN                    NaN   
3                   NaN                   NaN                    NaN   
4                   NaN                   NaN                    NaN   

  adwordsClickInfo.adNetworkType adwordsClickInfo.isVideoAd adContent  \
0                            NaN                        NaN       NaN   
1                            NaN                        NaN       NaN   
2                            NaN                        NaN       NaN   
3                            NaN                        NaN       NaN   
4                            NaN                        NaN       NaN   

  campaignCode  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 55 columns]

In [12]:
#Viewing first 5 rows of the testing dataset
test_data.head()

channelGrouping      date        fullVisitorId  \
0  Organic Search  20171016  6167871330617112363   
1  Organic Search  20171016  0643697640977915618   
2  Organic Search  20171016  6059383810968229466   
3  Organic Search  20171016  2376720078563423631   
4  Organic Search  20171016  2314544520795440038   

                        sessionId  socialEngagementType     visitId  \
0  6167871330617112363_1508151024  Not Socially Engaged  1508151024   
1  0643697640977915618_1508175522  Not Socially Engaged  1508175522   
2  6059383810968229466_1508143220  Not Socially Engaged  1508143220   
3  2376720078563423631_1508193530  Not Socially Engaged  1508193530   
4  2314544520795440038_1508217442  Not Socially Engaged  1508217442   

   visitNumber  visitStartTime continent      subContinent  ...  \
0            2      1508151024      Asia    Southeast Asia  ...   
1            1      1508175522    Europe   Southern Europe  ...   
2            1      1508143220    Europe    Western Europe  ...   
3            1      1508193530  Americas  Northern America  ...   
4            1      1508217442  Americas  Northern America  ...   

          keyword isTrueDirect adwordsClickInfo.criteriaParameters  \
0  (not provided)         True       not available in demo dataset   
1  (not provided)          NaN       not available in demo dataset   
2  (not provided)          NaN       not available in demo dataset   
3  (not provided)          NaN       not available in demo dataset   
4  (not provided)          NaN       not available in demo dataset   

  adwordsClickInfo.page adwordsClickInfo.slot adwordsClickInfo.gclId  \
0                   NaN                   NaN                    NaN   
1                   NaN                   NaN                    NaN   
2                   NaN                   NaN                    NaN   
3                   NaN                   NaN                    NaN   
4                   NaN                   NaN                    NaN   

  adwordsClickInfo.adNetworkType adwordsClickInfo.isVideoAd referralPath  \
0                            NaN                        NaN          NaN   
1                            NaN                        NaN          NaN   
2                            NaN                        NaN          NaN   
3                            NaN                        NaN          NaN   
4                            NaN                        NaN          NaN   

  adContent  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN  

[5 rows x 53 columns]

In [35]:
# date to YYYYMMDD
train_data['date'] = pd.to_datetime(train_data['date'], format='%Y%m%d')
test_data['date'] = pd.to_datetime(test_data['date'], format='%Y%m%d')

In [38]:
#check are there any date gaps in the dataset
print(train_data['date'].min(), train_data['date'].max())
print(test_data['date'].min(), test_data['date'].max())

# check the gap in the date
print(train_data['date'].max() - train_data['date'].min())
print(test_data['date'].max() - test_data['date'].min())

2016-08-01 00:00:00 2017-08-01 00:00:00
2017-08-02 00:00:00 2018-04-30 00:00:00
365 days 00:00:00
271 days 00:00:00


### From the above result we can see there are 365 of training dataset, the test dataset is 271 days, now is the time to test whether the time frame has a gap

In [40]:
date_range = pd.date_range(start='2016-08-01', end='2017-08-01')
missing_dates = date_range[~date_range.isin(train_data['date'].unique())]
print('There are {} missing dates in the training dataset'.format(len(missing_dates)))


date_range = pd.date_range(start='2017-08-02', end='2018-04-30')
missing_dates = date_range[~date_range.isin(test_data['date'].unique())]
print('There are {} missing dates in the test dataset'.format(len(missing_dates)))


There are 0 missing dates in the training dataset
There are 0 missing dates in the test dataset


### Below code are from Abdul, I refined the dataframe and code above
* <b> Current df_train is train_data </b>
* <b> Current df_test is test_data </b>
* Added a line allow df_test and df_train to run

In [17]:
#importing the training set
#df_train=pd.read_csv("/Users/Abdul/Desktop/MMA/Enterprise Data Science/Revenue-Radar/Data/train_dejsonified.csv")

# #Importing the test set
#df_test=pd.read_csv("/Users/Abdul/Desktop/MMA/Enterprise Data Science/Revenue-Radar/Data/test_dejsonified.csv")

In [22]:
print('The dimensions of the training dataset are:', df_train.shape)
print('The dimensions of the testing dataset are:', df_test.shape)

The dimensions of the training dataset are: (903653, 55)
The dimensions of the testing dataset are: (804684, 53)


In [23]:
# Printing all the column names for exploration and understanding the dataset
df_train.columns.to_list()

['channelGrouping',
 'date',
 'fullVisitorId',
 'sessionId',
 'socialEngagementType',
 'visitId',
 'visitNumber',
 'visitStartTime',
 'continent',
 'subContinent',
 'country',
 'region',
 'metro',
 'city',
 'cityId',
 'networkDomain',
 'latitude',
 'longitude',
 'networkLocation',
 'browser',
 'browserVersion',
 'browserSize',
 'operatingSystem',
 'operatingSystemVersion',
 'isMobile',
 'mobileDeviceBranding',
 'mobileDeviceModel',
 'mobileInputSelector',
 'mobileDeviceInfo',
 'mobileDeviceMarketingName',
 'flashVersion',
 'language',
 'screenColors',
 'screenResolution',
 'deviceCategory',
 'visits',
 'hits',
 'pageviews',
 'bounces',
 'newVisits',
 'transactionRevenue',
 'campaign',
 'source',
 'medium',
 'keyword',
 'adwordsClickInfo.criteriaParameters',
 'isTrueDirect',
 'referralPath',
 'adwordsClickInfo.page',
 'adwordsClickInfo.slot',
 'adwordsClickInfo.gclId',
 'adwordsClickInfo.adNetworkType',
 'adwordsClickInfo.isVideoAd',
 'adContent',
 'campaignCode']

In [24]:
# printing all the column names for exploration and understanding the dataset. Also, there are less columns in the test dataset. Upon inspection, it is found that the test dataset does not have the 'revenue' column which is the target variable that we are trying to predict.
df_test.columns.to_list()

['channelGrouping',
 'date',
 'fullVisitorId',
 'sessionId',
 'socialEngagementType',
 'visitId',
 'visitNumber',
 'visitStartTime',
 'continent',
 'subContinent',
 'country',
 'region',
 'metro',
 'city',
 'cityId',
 'networkDomain',
 'latitude',
 'longitude',
 'networkLocation',
 'browser',
 'browserVersion',
 'browserSize',
 'operatingSystem',
 'operatingSystemVersion',
 'isMobile',
 'mobileDeviceBranding',
 'mobileDeviceModel',
 'mobileInputSelector',
 'mobileDeviceInfo',
 'mobileDeviceMarketingName',
 'flashVersion',
 'language',
 'screenColors',
 'screenResolution',
 'deviceCategory',
 'visits',
 'hits',
 'pageviews',
 'newVisits',
 'bounces',
 'campaign',
 'source',
 'medium',
 'keyword',
 'isTrueDirect',
 'adwordsClickInfo.criteriaParameters',
 'adwordsClickInfo.page',
 'adwordsClickInfo.slot',
 'adwordsClickInfo.gclId',
 'adwordsClickInfo.adNetworkType',
 'adwordsClickInfo.isVideoAd',
 'referralPath',
 'adContent']

In [25]:
# Inspecting proportion of missing values in the transactioRevenue column (This info is useful for the classification model)
print('The proportion of missing values in the transactionRevenue column is:',df_train['transactionRevenue'].isnull().sum()/len(df_train))
# Inspecting the number of non-null values in the transactioRevenue column ( The subset is useful for the regression model)
print('The number of non-null values in the transactionRevenue column is:',df_train['transactionRevenue'].notnull().sum())

The proportion of missing values in the transactionRevenue column is: 0.9872572768529513
The number of non-null values in the transactionRevenue column is: 11515


In [26]:
def subset_data(df):
    reg_df=df[df['transactionRevenue'].notnull()]

    return reg_df


In [27]:
reg_df=subset_data(df_train)

In [28]:
reg_df.shape

(11515, 55)

In [29]:
reg_df.head()

channelGrouping       date        fullVisitorId  \
752          Direct 2016-09-02  6194193421514403509   
753  Organic Search 2016-09-02  5327166854580374902   
799        Referral 2016-09-02  8885051388942907862   
802        Referral 2016-09-02  0185467632009737931   
859        Referral 2016-09-02  3244885836845029978   

                          sessionId  socialEngagementType     visitId  \
752  6194193421514403509_1472843572  Not Socially Engaged  1472843572   
753  5327166854580374902_1472844906  Not Socially Engaged  1472844906   
799  8885051388942907862_1472827393  Not Socially Engaged  1472827393   
802  0185467632009737931_1472846398  Not Socially Engaged  1472846398   
859  3244885836845029978_1472824817  Not Socially Engaged  1472824817   

     visitNumber  visitStartTime continent      subContinent  ...  \
752            1      1472843572  Americas  Northern America  ...   
753            3      1472844906  Americas  Northern America  ...   
799            7      1472827393  Americas  Northern America  ...   
802            6      1472846398  Americas  Northern America  ...   
859            4      1472824817  Americas  Northern America  ...   

    adwordsClickInfo.criteriaParameters isTrueDirect referralPath  \
752       not available in demo dataset         True          NaN   
753       not available in demo dataset         True          NaN   
799       not available in demo dataset         True            /   
802       not available in demo dataset         True            /   
859       not available in demo dataset         True            /   

    adwordsClickInfo.page adwordsClickInfo.slot adwordsClickInfo.gclId  \
752                   NaN                   NaN                    NaN   
753                   NaN                   NaN                    NaN   
799                   NaN                   NaN                    NaN   
802                   NaN                   NaN                    NaN   
859                   NaN                   NaN                    NaN   

    adwordsClickInfo.adNetworkType adwordsClickInfo.isVideoAd adContent  \
752                            NaN                        NaN       NaN   
753                            NaN                        NaN       NaN   
799                            NaN                        NaN       NaN   
802                            NaN                        NaN       NaN   
859                            NaN                        NaN       NaN   

    campaignCode  
752          NaN  
753          NaN  
799          NaN  
802          NaN  
859          NaN  

[5 rows x 55 columns]

In [32]:
# count missing value
reg_df.isnull().sum()/len(reg_df)

channelGrouping                        0.000000
date                                   0.000000
fullVisitorId                          0.000000
sessionId                              0.000000
socialEngagementType                   0.000000
visitId                                0.000000
visitNumber                            0.000000
visitStartTime                         0.000000
continent                              0.000000
subContinent                           0.000000
country                                0.000000
region                                 0.000000
metro                                  0.000000
city                                   0.000000
cityId                                 0.000000
networkDomain                          0.000000
latitude                               0.000000
longitude                              0.000000
networkLocation                        0.000000
browser                                0.000000
browserVersion                         0

### The feature that missing too much data: 
* adwordsClickInfo.page                  0.960573
* adwordsClickInfo.slot                  0.960573
* adwordsClickInfo.gclId                 0.960399
* adwordsClickInfo.adNetworkType         0.960573
* adwordsClickInfo.isVideoAd             0.960573
* adContent                              0.987581
* campaignCode                           1.000000

###  below code is for saving the pre processed regression dataset

In [34]:
# Replace my path with yours
reg_df.to_csv('Data/reg_df.csv', index=False)